pybigquery/sqlalchemy_bigquery.py

```diff
@@ -156,19 +145,68 @@ def create_cursor(self):
             c.arraysize = self.dialect.arraysize
         return c
 
+    def get_insert_default(self, column):  # pragma: NO COVER
+        # Only used by compliance tests
+        if isinstance(column.type, Integer):
+            return random.randint(-9223372036854775808, 9223372036854775808)  # 1<<63
+        elif isinstance(column.type, String):
+            return str(uuid.uuid4())
+
+    def pre_exec(
+        self,
+        in_sub=re.compile(
+            r" IN UNNEST\(\[ "
+            r"(%\([^)]+\d+\)s(, %\([^)]+_\d+\)s)+)?"  # Placeholders
+            ":([A-Z0-9]+)"  # Type
+            r" \]\)"
```

Comment:
> Also, what digits are we trying to match?
> 
> ```
> (%\([^)]+\d+\)s
> ```
> 
> Are these parameter indexes SQLAlchemy puts in these placeholders?
> 
> I also noticed that in the first part, that "numeric" bit is not prefixed by an underscore, while in the second (optional) part it is: `_\d+` - on purpose or a typo?

Yes, you're write, on what they're bout and the typo.  I added the missing `_` as well as this comment.

```
        # In the placeholder part of the regex above, the `_\d+
        # suffixes refect that when an array parameter is expanded,
        # numeric suffixes are added.  For example, a placeholder like
        # `%(foo)s` gets expaneded to `%(foo_0)s, `%(foo_1)s, ...`.
```

Reply:
> > (nit)
> > I find it easier to read such strings if written as multiline string literals, because no opening and closing quotes in the pattern to distract the eyes. :) And in `re.VERBOSE` mode whitespace is already ignored, allowing for useful comments. The downside is that that those few space characters will need to be explicitly escaped, though.
> > Feel free to ignore if your brain is wired differently, or if you don't see enough value this change would bring.
> 
> I don't fully follow your comment.
> 
> How would you have written this?

I tried this.
- I found the space escaping distracting.
- It made the commends harder to see. This may be a personal (editor) problem.
- I trues 3 times and wasn't able to not break the regex. I must have been escaping the spaces wrong.

<shrug>

Reply:
Fair enough, let's leave it as is then. I personally found the verbose mode useful when writing some very complex replacement rules in `synth.py`, but having to explicitly escape any expected whitespace can indeed be a significant drawback.